File: go/day21_part2_2024.go
package main

import (
	"fmt"
	"os"
	"strings"
)

type Position struct {
	i, j int
}

type CacheKey struct {
	pos      Position
	ch       byte
	padType  string
}

type SolveKey struct {
	code      string
	robots    int
	maxRobots int
}

// Cache maps
var positionCache = make(map[string]Position)
var okCache = make(map[string]bool)
var moveCache = make(map[CacheKey]string)
var solveCache = make(map[SolveKey]int)

func findPosition(mat []string, ch byte) Position {
	// Check cache first
	key := string(ch) + strings.Join(mat, "")
	if pos, exists := positionCache[key]; exists {
		return pos
	}

	for i := range mat {
		for j := range mat[i] {
			if mat[i][j] == ch {
				pos := Position{i, j}
				positionCache[key] = pos
				return pos
			}
		}
	}
	return Position{-1, -1}
}

func ok(mat []string, st Position, seq string) bool {
	key := fmt.Sprintf("%d,%d,%s,%s", st.i, st.j, seq, strings.Join(mat, ""))
	if result, exists := okCache[key]; exists {
		return result
	}

	curr := st
	for i := 0; i < len(seq); i++ {
		if mat[curr.i][curr.j] == ' ' {
			okCache[key] = false
			return false
		}
		
		ch := seq[i]
		switch ch {
		case '^':
			curr.i--
		case 'v':
			curr.i++
		case '<':
			curr.j--
		case '>':
			curr.j++
		}
		
		if curr.i < 0 || curr.i >= len(mat) || curr.j < 0 || curr.j >= len(mat[0]) {
			okCache[key] = false
			return false
		}
	}
	
	okCache[key] = true
	return true
}

func generateMoves(position Position, objective byte, pad []string) string {
	key := CacheKey{position, objective, strings.Join(pad, "")}
	if moves, exists := moveCache[key]; exists {
		return moves
	}

	objPos := findPosition(pad, objective)
	
	ret := strings.Builder{}
	// First attempt
	if position.j > objPos.j {
		ret.WriteString(strings.Repeat("<", position.j-objPos.j))
	}
	if position.i > objPos.i {
		ret.WriteString(strings.Repeat("^", position.i-objPos.i))
	}
	if position.i < objPos.i {
		ret.WriteString(strings.Repeat("v", objPos.i-position.i))
	}
	if position.j < objPos.j {
		ret.WriteString(strings.Repeat(">", objPos.j-position.j))
	}

	result := ret.String()
	if !ok(pad, position, result) {
		ret.Reset()
		if position.j < objPos.j {
			ret.WriteString(strings.Repeat(">", objPos.j-position.j))
		}
		if position.i > objPos.i {
			ret.WriteString(strings.Repeat("^", position.i-objPos.i))
		}
		if position.i < objPos.i {
			ret.WriteString(strings.Repeat("v", objPos.i-position.i))
		}
		if position.j > objPos.j {
			ret.WriteString(strings.Repeat("<", position.j-objPos.j))
		}
		result = ret.String()
	}
	
	moveCache[key] = result
	return result
}

func solve(code string, robots int, keyPad []string, robotPad []string, maxRobots int) int {
	key := SolveKey{code, robots, maxRobots}
	if result, exists := solveCache[key]; exists {
		return result
	}

	if robots <= 0 {
		return len(code)
	}

	ret := 0
	posi, posj := 3, 2
	if robots != maxRobots {
		posi = 0
	}

	var moves string
	for i := 0; i < len(code); i++ {
		ch := code[i]
		if robots == maxRobots {
			moves = generateMoves(Position{posi, posj}, ch, keyPad)
			pos := findPosition(keyPad, ch)
			posi, posj = pos.i, pos.j
		} else {
			moves = generateMoves(Position{posi, posj}, ch, robotPad)
			pos := findPosition(robotPad, ch)
			posi, posj = pos.i, pos.j
		}
		ret += solve(moves+"A", robots-1, keyPad, robotPad, maxRobots)
	}

	solveCache[key] = ret
	return ret
}

func main() {
	content, err := os.ReadFile("input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}

	maxRobots := 26 // 25 robots with directional keypads + 1 robot with numeric keypad
	keyPad := []string{
		"789",
		"456",
		"123",
		" 0A",
	}
	robotPad := []string{
		" ^A",
		"<v>",
	}

	ret := 0
	codes := strings.Split(strings.TrimSpace(string(content)), "\n")
	
	for _, code := range codes {
		code = strings.TrimSpace(code)
		if code == "" {
			continue
		}

		numericPart := 0
		for i := 0; i < len(code); i++ {
			if code[i] >= '0' && code[i] <= '9' {
				numericPart = numericPart*10 + int(code[i]-'0')
			}
		}

		sv := solve(code, maxRobots, keyPad, robotPad, maxRobots)
		ret += sv * numericPart
	}

	fmt.Println(ret)
}

File: nim/day21_part2_2024.nim

import strutils, sequtils, tables, os

type
  Position = tuple[i, j: int]
  CacheKey = tuple[pos: Position, ch: char, padType: string]
  SolveKey = tuple[code: string, robots, maxRobots: int]

var
  positionCache: Table[string, Position]
  okCache: Table[string, bool]
  moveCache: Table[CacheKey, string]
  solveCache: Table[SolveKey, int]

proc findPosition(mat: seq[string], ch: char): Position =
  let key = $ch & mat.join()
  if key in positionCache:
    return positionCache[key]
  for i, row in mat:
    for j, c in row:
      if c == ch:
        let pos = (i, j)
        positionCache[key] = pos
        return pos
  return (-1, -1)

proc ok(mat: seq[string], st: Position, seq: string): bool =
  let key = $st.i & "," & $st.j & "," & seq & "," & mat.join()
  if key in okCache:
    return okCache[key]
  var curr = st
  for ch in seq:
    if mat[curr.i][curr.j] == ' ':
      okCache[key] = false
      return false
    case ch
    of '^':
      curr.i -= 1
    of 'v':
      curr.i += 1
    of '<':
      curr.j -= 1
    of '>':
      curr.j += 1
    else:
      discard
    if curr.i < 0 or curr.i >= mat.len or curr.j < 0 or curr.j >= mat[0].len:
      okCache[key] = false
      return false
  okCache[key] = true
  return true

proc generateMoves(position: Position, objective: char, pad: seq[string]): string =
  let key = (position, objective, pad.join())
  if key in moveCache:
    return moveCache[key]
  let objPos = findPosition(pad, objective)
  var ret = ""
  if position.j > objPos.j:
    ret &= repeat("<", position.j - objPos.j)
  if position.i > objPos.i:
    ret &= repeat("^", position.i - objPos.i)
  if position.i < objPos.i:
    ret &= repeat("v", objPos.i - position.i)
  if position.j < objPos.j:
    ret &= repeat(">", objPos.j - position.j)
  var result = ret
  if not ok(pad, position, result):
    ret = ""
    if position.j < objPos.j:
      ret &= repeat(">", objPos.j - position.j)
    if position.i > objPos.i:
      ret &= repeat("^", position.i - objPos.i)
    if position.i < objPos.i:
      ret &= repeat("v", objPos.i - position.i)
    if position.j > objPos.j:
      ret &= repeat("<", position.j - objPos.j)
    result = ret
  moveCache[key] = result
  return result

proc solve(code: string, robots: int, keyPad: seq[string], robotPad: seq[string], maxRobots: int): int =
  let key = (code, robots, maxRobots)
  if key in solveCache:
    return solveCache[key]
  if robots <= 0:
    return code.len
  var ret = 0
  var posi = 3
  var posj = 2
  if robots != maxRobots:
    posi = 0
  for ch in code:
    var moves: string
    if robots == maxRobots:
      moves = generateMoves((posi, posj), ch, keyPad)
      let pos = findPosition(keyPad, ch)
      posi = pos.i
      posj = pos.j
    else:
      moves = generateMoves((posi, posj), ch, robotPad)
      let pos = findPosition(robotPad, ch)
      posi = pos.i
      posj = pos.j
    ret += solve(moves & "A", robots - 1, keyPad, robotPad, maxRobots)
  solveCache[key] = ret
  return ret

let
  maxRobots = 26
  keyPad = @[
    "789",
    "456",
    "123",
    " 0A"
  ]
  robotPad = @[
    " ^A",
    "<v>"
  ]

var ret = 0
let content = readFile("input.txt").strip().splitLines()
for code in content:
  let code = code.strip()
  if code.len == 0:
    continue
  var numericPart = 0
  for ch in code:
    if ch >= '0' and ch <= '9':
      numericPart = numericPart * 10 + (ord(ch) - ord('0'))
  ret += solve(code, maxRobots, keyPad, robotPad, maxRobots) * numericPart
echo ret
